elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore.rb

```diff
@@ -43,97 +46,156 @@ module HTTP
+            with_request_retries do
+              body = __convert_to_json(body) if body
+              enriching_response(method, path, params, body) do
+                url, response = @pool.perform_request(method, path, params, body)
 
-            if options.key?(:headers)
-              @request_options[:headers] = options[:headers]
+                # Raise an exception so we can catch it for `retry_on_status`
+                __raise_transport_error(response) if response.status.to_i >= 300 && @retry_on_status.include?(response.status.to_i)
+                [url, response]
+              end
             end
+          end
 
-            Connections::Collection.new \
-              :connections => hosts.map { |host|
-                host[:protocol]   = host[:scheme] || DEFAULT_PROTOCOL
-                host[:port]     ||= DEFAULT_PORT
+          # This takes a host string to aid in debug logging
+          def with_request_retries
+            tries = 0
+            begin
+              tries += 1
+              yield
+            rescue ::Elasticsearch::Transport::Transport::ServerError => e
+              if @retry_on_status.include?(e.response.status)
+                logger.warn "[#{e.class}] Attempt #{tries} to get response from #{url}" if logger
+                logger.debug "[#{e.class}] Attempt #{tries} to get response from #{url}" if logger
+                if tries <= max_retries
+                  retry
+                else
+                  logger.fatal "[#{e.class}] Cannot get response from #{url} after #{tries} tries" if logger
+                  raise e
+                end
+              else
+                raise e
+              end
+            rescue ::Elasticsearch::Transport::Transport::HostUnreachableError => e
+              logger.error "[#{e.class}] #{e.message} #{e.url}" if logger
 
-                host.delete(:user)     # auth is not supported here.
-                host.delete(:password) # use the headers
+              if @options[:retry_on_failure]
+                logger.warn "[#{e.class}] Attempt #{tries} connecting to #{connection.host.inspect}" if logger
+                if tries <= max_retries
+                  if @options[:reload_on_failure] && pool.alive_urls_count == 0
+                    logger.warn "[#{e.class}] Reloading connections (attempt #{tries} of #{connections.all.size})" if logger
+                    reload_connections!
+                  end
 
-                Connections::Connection.new \
-                  :host => host,
-                  :connection => @manticore
-              },
-              :selector_class => options[:selector_class],
-              :selector => options[:selector]
+                  retry
+                else
+                  logger.fatal "[#{e.class}] Cannot connect to #{connection.host.inspect} after #{tries} tries" if logger
+                  raise e
+                end
+              end
+            rescue Exception => e
```

Comment:
We re-raise it anyway. I'm always worried about other exceptions popping up on JRuby. WDYT?